python/ribasim/ribasim/model.py

```diff
@@ -280,6 +285,137 @@ def write(self, filepath: str | PathLike[str]) -> Path:
+            raise ValueError("Minimum control inneighbor or outneighbor unsatisfied")
+
+    def _check_neighbor_amount(
+        self,
+        df_graph: pd.DataFrame,
+        edge_amount: dict[str, list[int]],
+        edge_type: str,
+        nodes,
+    ) -> bool:
+        """Check if the neighbor amount of the two nodes connected by the given edge meet the minimum requirements."""
+
+        is_valid = True
+
+        # filter graph by edge type
+        df_graph = df_graph.loc[df_graph["edge_type"] == edge_type]
+
+        # count occurrence of "from_node" which reflects the number of outneighbors
+        from_node_count = (
+            df_graph.groupby("from_node_id").size().reset_index(name="from_node_count")  # type: ignore
+        )
+
+        # append from_node_count column to from_node_id and from_node_type
+        from_node_info = (
+            df_graph[["from_node_id", "from_node_type"]]
+            .drop_duplicates()
+            .merge(from_node_count, on="from_node_id", how="left")
+        )
+        from_node_info = from_node_info[
+            ["from_node_id", "from_node_count", "from_node_type"]
+        ]
+
+        # add the node that is not the upstream of any other nodes
+        from_node_info = self._add_source_sink_node(nodes, from_node_info, "from")
+
+        # loop over all the "from_node" and check if they have enough outneighbor
+        for _, row in from_node_info.iterrows():
+            # from node's outneighbor
+            if row["from_node_count"] < edge_amount[row["from_node_type"]][2]:
+                is_valid = False
+                logging.error(
+                    f"Node {row['from_node_id']} must have at least {edge_amount[row['from_node_type']][2]} outneighbor(s) (got {row['from_node_count']})"
+                )
+
+        # count occurrence of "to_node" which reflects the number of inneighbors
+        to_node_count = (
+            df_graph.groupby("to_node_id").size().reset_index(name="to_node_count")  # type: ignore
+        )
+
+        # append to_node_count column to result
+        to_node_info = (
+            df_graph[["to_node_id", "to_node_type"]]
+            .drop_duplicates()
+            .merge(to_node_count, on="to_node_id", how="left")
+        )
+        to_node_info = to_node_info[["to_node_id", "to_node_count", "to_node_type"]]
+
+        # add the node that is not the downstream of any other nodes
+        to_node_info = self._add_source_sink_node(nodes, to_node_info, "to")
+
+        # loop over all the "to_node" and check if they have enough inneighbor
+        for _, row in to_node_info.iterrows():
+            if row["to_node_count"] < edge_amount[row["to_node_type"]][0]:
+                is_valid = False
+                logging.error(
+                    f"Node {row['to_node_id']} must have at least {edge_amount[row['to_node_type']][0]} inneighbor(s) (got {row['to_node_count']})"
+                )
+
+        return is_valid
+
+    def _add_source_sink_node(
```

Comment:
Not sure what source_sink means here? I'd suggest something like `_get_isolated_nodes`?

Reply:
In graph theory "source" is a node that doesn't have "in" edge and "sink" is a node that doesn't have "out" edge.

Maybe we have some more comprehensive names for these type of nodes?

Reply:
Well sure, but something that's both a source and a sink, is neither. A source or a sink would be present in the edge table. You're looking for vertices with degree zero right? Those are often called isolate(d) (see for example this method https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.isolate.number_of_isolates.html), hence my suggestion for `_get_isolated_nodes`.

Reply:
The main purpose of this function is to check the nodes that don't have one of the in- or out edge. In every model there are nodes like this. If a node doesn't have an in edge, it will not be checked because it doesn't end up in the to_node_count column. Same for the nodes that don't have out edge and don't appear in from_node_count column. So this function adds those nodes and specifies that there is 0 in/out edge.

This function can handle isolated node as well. However the case of isolated node is much less common.

Reply:
Already did ^o^